Internal #330: Quantile Performance

hawkfish opened the pull request:
Rewrite of moving `quantile` and `mad` to use better/faster data structures.

* Add new `window_init` aggregate API to allow custom window functions to create global data structures
* Compute and pass down statistics for frame computations to support global vs local performance optimisations
* Add third party skip list code for thread-local implementations
* Add (rewritten) third party merge sort tree code for global implementations
* Extend MST `SelectNth` method to handle new `EXCLUDE` functionality
* Fix problem with `ValidityMask::Resize` where it was always allocating memory instead of deferring if all rows were valid

fixes: duckdblabs/duckdb-internal#330

Commit:
Internal #330: Import Skip Lists

Add Python implementation of skip lists to libraries.

Commit:
Merge branch 'main' into window-median

Commit:
Internal #330: Quantile Skip Lists

Convert basic scalar quantile window to use skip lists.
Performance on 1e9 rows, single partition, 16 cores
went from 11:11 to 00:36.

Commit:
Internal #330: Quantile Skip Lists

Convert scalar and list windowed quantile to use skip lists.
Performance on 1e9 rows, single partition, 16 cores
went from 11:11 to 00:36.

Commit:
Merge branch 'quantile-value' into window-median

Commit:
Merge branch 'main' into window-median

Commit:
Internal #330: Merge Sort Trees

Create MST template for use in quantile and other aggregates.

Commit:
Internal #330: Window Global Init

* Define QuantileSortTree for quantile acceleration.
* Add aggregate_wininit_t for constructing global windowing state.
* Update aggregate_window_t to take optional global windowing state.
* Add quantile WindowInit function (not used yet).

Commit:
Merge branch 'feature' into merge-sort-trees

Commit:
Internal #330: Window Global Init

Call window initialisation if it is defined.

hawkfish committed:
Merge branch 'main' into merge-sort-trees

hawkfish committed:
Internal #330: Quantile Merge Sort Trees

Switch quantile computations to use merge sort trees.
MAD not converted yet.

hawkfish committed:
Internal #330: Quantile Merge Sort Trees

Fix double decrement.

hawkfish committed:
Internal #330: Quantile Merge Sort Trees

Add MIT License.

hawkfish committed:
Internal #330: Quantile Merge Sort Trees

Toggle 32/64 bit versions.

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Internal #330: MAD Merge Sort Trees

Refactor, simplify and deploy to Median Absolute Deviation.

hawkfish committed:
Merge branch 'feature' into window-median

hawkfish committed:
Merge branch 'window-median' into merge-sort-trees

hawkfish committed:
Internal #330: Window Frame Statistics

Plumb through statistics and extract min/max for ROWS.

hawkfish committed:
Internal #330: Quantile Statistics

Use frame expression statistics to choose between
Skip Lists and Merge Sort Trees

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Internal #330: Window EXCLUDE Statistics

Test statistics and handle no propagation situation.

hawkfish committed:
Internal #330: Merge Sort Trees

Fix fanout misplacement.

hawkfish committed:
Internal #330: Merge Sort Tree EXCLUDE

Finish implementing and testing merge sort trees
by supporting EXCLUDE/multiple frames.

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Internal #330: Merge Sort Trees

Fix CI issues (formatting, includes)

hawkfish committed:
Internal #330: Import Skip Lists

Remove non-portable #pragma mark lines.

hawkfish committed:
Internal #330: Merge Sort Tree EXCLUDE

Standardise SubFrame typing.

hawkfish committed:
Internal #330: Quantile EXCLUDE

Push subframes into the Merge Sort Tree.

hawkfish committed:
Merge branch 'main' into merge-sort-trees

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Internal #330: Validity Mask Hack

Temporary hack around unnecessary validity mask construction in DataChunk::Append.

hawkfish committed:
Internal #330: Validity Mask Fix

Try just fixing the validity mask allocation.

hawkfish committed:
Internal #330: Validity Mask Fix

Find remaining shubs and zuuls.

hawkfish commented:
Current performance:

![image](https://github.com/duckdb/duckdb/assets/13156216/f1c471ef-619f-4c03-951a-90ae7658a993)

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Internal #330: Skip List Pooling

Add a one element Node pool to avoid repeated new/delete calls.

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish commented:
The one element pool gives a 10% boost on the original benchmark. Given that 50% of the time is sorting the data (which pandas is avoiding), this is significant.

![image](https://github.com/duckdb/duckdb/assets/13156216/03c71e12-eae2-43d3-b042-07af4387a823)

hawkfish committed:
Merge branch 'main' into merge-sort-trees

taniabogatsch commented on src/include/duckdb/common/vector_operations/aggregate_executor.hpp:
```diff
@@ -17,13 +17,18 @@ namespace duckdb {
 };
 
+// A set of window subframes for windowed EXCLUDE
+using SubFrames = vector<FrameBounds>;
```
Before looking at this file I was wondering about the name `SubFrames`. Can we rename this to something like `ExcludedFrames`?

hawkfish replied:
Well they are actually _included_ frames, so that's probably not a good idea. I used to call them excluded frames but Adrian and I started calling them "subframes" so I think that is now the term of art(!)

taniabogatsch replied:
Okay makes sense, then!

taniabogatsch commented on src/include/duckdb/function/aggregate_function.hpp:
```diff
@@ -15,6 +15,14 @@
 namespace duckdb {
 
+//! The half-open range of frame boundary values relative to the current row
+struct FrameStats {
```
I wonder if this is the best place to put this struct. From my understanding, frames are a concept that is very specific to the windowing functionality, and not to aggregation in general.

hawkfish replied:
It put it there because the windowing APIs use it and they are defined in that file. Prying apart window and general aggregation is unfortunately rather difficult.

taniabogatsch commented on src/include/duckdb/function/aggregate_function.hpp:
```diff
@@ -40,10 +48,15 @@ typedef void (*aggregate_destructor_t)(Vector &state, AggregateInputData &aggr_i
 typedef void (*aggregate_window_t)(Vector inputs[], const ValidityMask &filter_mask,
                                    AggregateInputData &aggr_input_data, idx_t input_count, data_ptr_t state,
-                                   const vector<FrameBounds> &frames, Vector &result, idx_t rid);
+                                   const SubFrames &frames, Vector &result, idx_t rid, const_data_ptr_t win_state);
```
Maybe we should be more explicit? `data_ptr_t` is also a window state (`win_state`). What about `l_state` and `g_state`? I think that this convention is somewhat common for other parts of duckdb code.

hawkfish replied:
Good call - I started doing that in other places but never went back to the headers.

taniabogatsch commented on src/execution/window_executor.cpp:
```diff
@@ -3,6 +3,8 @@
 #include "duckdb/common/operator/add.hpp"
 #include "duckdb/common/operator/subtract.hpp"
 
+#include <array>
```
Nit: `duckdb/common/array.hpp`

taniabogatsch reviewed (changes requested):
Hi @hawkfish, cool PR! I added a few comments.

I didn't have an in-depth look at `src/core_functions/aggregate/holistic/quantile.cpp` and `src/include/duckdb/execution/merge_sort_tree.hpp` yet. I can do that at a later point this week. I assume that I do not have to look at the skip list code?

Something that came to my mind is that this PR does not include any benchmarks. I think we should add at least a few, to avoid future regressions. Or did I miss them, and they already exist?

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish commented on src/include/duckdb/execution/merge_sort_tree.hpp:
I may go back one day and tweak some stuff for the 32/64 bit memory but I think these are all OK to change.

hawkfish commented:
> Hi @hawkfish, cool PR! I added a few comments.
> 
> I didn't have an in-depth look at `src/core_functions/aggregate/holistic/quantile.cpp` and `src/include/duckdb/execution/merge_sort_tree.hpp` yet. I can do that at a later point this week. I assume that I do not have to look at the skip list code?
> 
> Something that came to my mind is that this PR does not include any benchmarks. I think we should add at least a few, to avoid future regressions. Or did I miss them, and they already exist?

The benchmarks in the visualisations are existing benchmarks (bottom 4) or ones planned for the internal db-benchmark suite (top 3). I'd move the top three into the suite but I'd have to get the data from @TMonster and I don't know if we want that kind of duplication?

hawkfish commented:
> I didn't have an in-depth look at `src/core_functions/aggregate/holistic/quantile.cpp` and `src/include/duckdb/execution/merge_sort_tree.hpp` yet. I can do that at a later point this week. I assume that I do not have to look at the skip list code?

The merge sort tree code is Adrian's but I modified it so we could compile it and we collaborated on adding the `EXCLUDE` changes. I also removed all his messy debugging code and added some whitespace.

The skip list code is a direct import of the code used by Python (it is a separate project). I just added a very simple one element memory pool (pooling is on their TODO list.) and you can look at the diff there (just routing new and delete through an internal class.)

hawkfish committed:
Internal #330: Quantile Performance Review

Integrate Tania' feedback (thanks!)

carlopi commented on third_party/skiplist/SkipList.cpp:
```diff
@@ -0,0 +1,90 @@
+ * @return rand() < RAND_MAX - RAND_MAX / 16;
+ * @endcode
+ */
+bool tossCoin() {
+    return rand() < RAND_MAX / 2;
+}
+
+void seedRand(unsigned seed) {
+    srand(seed);
+}
```
This seems a strange choice, since it's somehow wasteful to compute N bits of randomness to use only 1 of them. But I don't think this is ever relevant for real.
Slightly more problematic, this messes with the state of the random generator, that I would probably avoid (`seedRand` is never actually called luckily), plus I think this might forces synchronisation cross threads, while here we are fine with thread local not-really random sources.

I would have gone with:
```
std::independent_bits_engine<std::default_random_engine, /*bits*/ 1, int> tossCoin;
if (tossCoin())  {
   // yey!
} else {
    // noo
}
```

(I went a bit in-depth on a not really relevant part)

hawkfish replied:
Interesting observation. I suppose since pandas is single-threaded this was not an issue for them. I suppose I could stuff in a thread-safe prng (I think we have one?)

hawkfish replied:
Reverting this because it trashes performance.

carlopi commented on third_party/skiplist/SkipList.cpp:
```diff
@@ -0,0 +1,90 @@
+    oss << "Index out of range 0 <= index < " << index;
+    std::string err_msg = oss.str();
+#else
+void _throw_exceeds_size(size_t /* index */) {
+    std::string err_msg = "Index out of range.";
+#endif
+    throw IndexError(err_msg);
+}
```
Do we need to catch these IndexError / wrap them in InternalErrors? Not really sure when this might happen.

hawkfish replied:
It's basically a `D_ASSERT`, so I think we are OK. It only gets called in `const` methods so it really signals either a caller error or data corruption.

hawkfish replied:
There is only one place to wrap - easy enough.

carlopi reviewed (commented):
Cool results, I only took a look at the skiplist library, minor comments

hawkfish committed:
Internal #330: Quantile Performance Review

Incorporate Carlo's review (thanks!)

hawkfish committed:
Internal #330: Quantile Performance Review

Fix signed/unsigned portability issue and extra include.

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

hawkfish committed:
Internal #330: Quantile Performance Review

Revert single-bit prng use as it trashes performance by 2x.

Mytherin reviewed (commented):
Thanks for the PR! Some minor comments. I also launched a [NightlyTests](https://github.com/hawkfish/duckdb/actions/runs/6744169145) to check the code coverage on this feature. Perhaps some more tests could be added after that to ensure the new code is fully tested? I'm not sure how many tests we have that correctly test all of the edge cases here.

hawkfish commented on src/include/duckdb/execution/window_segment_tree.hpp:
```diff
@@ -44,9 +44,14 @@ class WindowAggregator {
 	//	Build
 	virtual void Sink(DataChunk &payload_chunk, SelectionVector *filter_sel, idx_t filtered);
-	virtual void Finalize();
+	virtual void Finalize(const FrameStats *stats);
```
It's actually a two-element array, so it was either this or `const array<FrameStats> &`. Would that be better?

hawkfish commented on src/execution/window_segment_tree.cpp:
```diff
@@ -308,8 +318,13 @@ void WindowCustomAggregator::Evaluate(WindowAggregatorState &lstate, const DataC
 	auto &lcstate = lstate.Cast<WindowCustomAggregatorState>();
 	auto &frames = lcstate.frames;
-	auto params = lcstate.inputs.data.data();
+	auto params = const_cast<DataChunk &>(inputs).data.data();
```
It is because of the `window` signature but since I'm fixing `window_init` I can deal with that too.

hawkfish replied:
Turns out there is a place in `ListWindow` that assumes it can munge the partition data, but now that the API has been fixed this is glaringly obvious...

hawkfish commented:
> Thanks for the PR! Some minor comments. I also launched a [NightlyTests](https://github.com/hawkfish/duckdb/actions/runs/6744169145) to check the code coverage on this feature. Perhaps some more tests could be added after that to ensure the new code is fully tested? I'm not sure how many tests we have that correctly test all of the edge cases here.

Yeah there is coverage in the benchmark suite but I could certainly move those queries into a `test_coverage` file now that they are faster...

Commit:
Internal #330: Custom Window APIs

Clean them up in response to PR feedback.
Note that ListWindow is being naughty and modifying the
partition data, which is not thread-safe.

Commit:
Internal #330: Quantile Performance Feedback

Restore the statistics API to use signed values.
Add coverage tests for quantile variants.

hawkfish committed:
Merge branch 'main' into merge-sort-trees

hawkfish committed:
Merge branch 'feature' into merge-sort-trees

Mytherin commented:
Thanks! LGTM
